yapapi/props/base.py

```diff
@@ -149,7 +187,156 @@ def __init__(self, iter):
+        default=default,
+        metadata={
+            PROP_KEY: key,
+            PROP_OPERATOR: operator,
+            PROP_MODEL_FIELD_TYPE: ModelFieldType.constraint,
+        },
+    )
+
+
+def prop(key: str, *, default=MISSING):
+    """
+    Return a property-type dataclass field for a Model.
+
+    :param key: the key of the property, e.g. "golem.runtime.name"
+    :param default: the default value of the property
+
+    example:
+    ```python
+    >>> from dataclasses import dataclass
+    >>> from yapapi.props.base import Model, prop
+    >>> from yapapi.props.builder import DemandBuilder
+    >>>
+    >>> @dataclass
+    ... class Foo(Model):
+    ...     max_bar: int = prop("bar", default=100)
+    ...
+    >>> builder = DemandBuilder()
+    >>> builder.add(Foo())
+    >>> print(builder.properties)
+    {'bar': 100}
+    ```
+    """
+    return field(
+        default=default, metadata={PROP_KEY: key, PROP_MODEL_FIELD_TYPE: ModelFieldType.property}
+    )
+
+
+def constraint_to_str(value, f: Field) -> str:
+    """Get string representation of a constraint with a given value.
+
+    :param value: the value of the the constraint field
+    :param f: the dataclass field for this constraint
+    """
+    return f"({f.metadata[PROP_KEY]}{f.metadata[PROP_OPERATOR]}{value})"
+
+
+def constraint_model_serialize(m: Model) -> List[str]:
+    """
+    Return a list of stringified constraints on the given Model instance.
+
+    :param m: instance of Model for which we want the constraints serialized
+    """
+    return [
+        constraint_to_str(getattr(m, f.name), f)
+        for f in fields(type(m))
+        if f.metadata.get(PROP_MODEL_FIELD_TYPE, "") == ModelFieldType.constraint
+    ]
+
+
+def join_str_constraints(constraints: List[str], operator: ConstraintGroupOperator = "&") -> str:
+    """
+    Join a list of constraints using the given opererator.
+
+    :param constraints: list of strings representing individual constraints
+                        (which may include previously joined constraint groups)
+    :param operator: constraint group operator, one of "&", "|", "!", which represent
+                     "and", "or" and "not" operations on those constraints.
+                     "!" requires that the list contains one and only one constraint.
+                     Defaults to "&" (and) if not given.
+    :return: string representation of the compound constraint.
+
+    example:
+    ```python
+    >>> from dataclasses import dataclass
+    >>> from yapapi.props.base import Model, constraint, constraint_model_serialize, join_str_constraints
+    >>>
+    >>> @dataclass
+    ... class Foo(Model):
+    ...     min_bar: int = constraint("bar", operator=">=", default=42)
+    ...     max_bar: int = constraint("bar", operator="<=", default=128)
+    ...
+    >>> print(join_str_constraints(constraint_model_serialize(Foo())))
+    (&(bar>=42)
+        (bar<=128))
+    ```
+    """
+    if not constraints:
+        return "()"
```

Comment:
Hmm, is the empty list of constraints a valid argument? If so, should we interpret `join_str_constraints([], "&")` and `join_str_constraints([], "|")` the same? Logically/algebraically, `&()` would be the empty list of constraints and it should be satisfied by all offers. Dually, `|()` is the empty list of alternatives and should be satisfied by no offer. And `join_str_constraints([], "!")` should probably be an error. Yet we return `"()"` in all cases.